## 6장. 상관관계분석

### 6.1 수치형 변수의 상관관계분석

#### 1. 데이터 불러오기

실습데이터 Ashopping을 파이썬을 이용하여 불러온다. 


#### 변수명=pd.read_csv("파일경로\파일이름\.파일형식", sep="구분자", encoding='코딩방식')

    구분자: ",",  "|", "\s+"
    encoding:'CP949' 파이썬에서 한글을 열수 있도록 인코딩 설정


In [1]:
# 텍스트 파일 열기
import pandas as pd
df = pd.read_csv("C:\python\Ashopping.txt",sep=",", encoding='CP949')

총 매출액과 방문빈도, 구매카테고리 수 간의 선형적인 연관성을 갖는지 확인해 보자. 

- 귀무가설: 총 매출액과 방문빈도, 구매 카테고리 수 간의 선형적인 관계가 존재하지 않는다. 
- 연구가설: 총 매출액과 방문빈도, 구매 카테고리 수 간의 선형적인 관계가 존재한다. 

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 42 columns):
고객ID                1000 non-null int64
이탈여부                1000 non-null int64
총_매출액               1000 non-null int64
방문빈도                1000 non-null int64
1회_평균매출액            1000 non-null int64
할인권_사용_횟수           1000 non-null int64
총_할인_금액             1000 non-null int64
고객등급                1000 non-null int64
구매유형                1000 non-null int64
클레임접수여부             1000 non-null int64
구매_카테고리_수           1000 non-null int64
거주지역                1000 non-null int64
성별                  1000 non-null int64
고객_나이대              1000 non-null int64
거래기간                1000 non-null int64
할인민감여부              1000 non-null int64
멤버쉽_프로그램_가입전_만족도    1000 non-null int64
멤버쉽_프로그램_가입후_만족도    1000 non-null int64
Recency             1000 non-null int64
Frequency           1000 non-null int64
Monetary            1000 non-null int64
상품_만족도              1000 non-null int64
매장_만족도        

In [5]:
df1=df[["총_매출액","방문빈도","구매_카테고리_수"]]

In [6]:
df1.head()

총_매출액  방문빈도  구매_카테고리_수
0   4007080    17          6
1   3168400    14          4
2   2680780    18          6
3   5946600    17          5
4  13745950    73          6

In [7]:
df1.corr(method='pearson')#수치형 상관관계 분석

총_매출액      방문빈도  구매_카테고리_수
총_매출액      1.000000  0.631171   0.308347
방문빈도       0.631171  1.000000   0.395861
구매_카테고리_수  0.308347  0.395861   1.000000

피어슨 상관관계 분석결과, 매출액과 방문빈도간의 상관관계가 매출액과 구매 카테고리 수와의 상관관계보다 높다고 할 수 있다. 따라서, A 쇼핑의 매출액을 강화시키기 위해서는 고객들의 구매 카테고리 수를 증가시키기 보다는 고객들의 더 잦은 방문을 유도하는 것이 필요하다고 볼 수 있다. 


### 6.2 편(부분) 상관관계분석

총 매출액과 방문빈도간의 상관관계가 구매 카테고리 수에 의해 영향을 받는지에 대해 실습

- 귀무가설 = 구매카테고리 수가  통제된 상황에서 총 매출액과 방문빈도 간에는 유의한 선형 관계를 갖지 않는다.
- 연구가설 = 구매카테고리 수가  통제된 상황에서 총 매출액과 방문빈도 간에는 유의한 선형 관계를 갖는다.

외부 패키지 설치

- conda install -c conda-forge pingouin

In [8]:
df1=df[["총_매출액","방문빈도","구매_카테고리_수"]]

In [9]:
from pingouin import partial_corr

In [10]:
partial_corr(data=df1, x='총_매출액', y='방문빈도', covar='구매_카테고리_수')

n      r         CI95%    r2  adj_r2         p-val  power
pearson  1000  0.583  [0.54, 0.62]  0.34   0.338  5.054646e-92    1.0

편(부분) 상관관계 분석결과 구매 카테고리 수를 통제한 상태에서 총 매출액과 방문빈도 간의 상관계수는 0.583 으로 도출되었다. 구매카테고리 수를 통제하지 않은 상태에서의 총 매출액과 방문빈도와의 상관계수인 0.631보다는 상당히 많이 그 정도가 약해졌다고 볼 수 있다. 따라서, 방문빈도가 고객들의 총 매출액에 많은 영향을 주는 변수임에는 분명하지만, 고객들의 구매다양성을 나타내는 구매 카테고리 수를 전혀 고려하지 않는다면 방문빈도와 매출액간의 관계는 생각보다 크지 않다는 것을 알 수 있다. 구매 카테고리 수가 매출액과의 직접적인 상관관계가 크지 않지만, 방문빈도와 매출액간의 관계에 영향을 미치므로 마케팅 전략 수립 시 함께 고민해야 할 필요가 있다.


### 6.3. 순서형 변수의 상관관계분석

#### 1회 평균 매출액 순위와 방문빈도 순위가 연관성이 있는지 확인해보자.

- 귀무가설: 1회 평균 매출액 순위와 방문빈도 순위는 연관성이 없다. 
- 연구가설: 1회 평균 매출액 순위와 방문빈도 순위는 연관성이 있다. 

In [11]:
df2=df[["1회_평균매출액","방문빈도"]]

In [12]:
df3 = df2.rank() #rank(): 기본적으로 값을 오름차순으로 순위를 매김. ascending=False 설정 내림차순 

In [13]:
df3.head()

1회_평균매출액   방문빈도
0     445.0  373.0
1     416.0  273.0
2     179.0  411.0
3     678.0  373.0
4     279.0  956.0

In [14]:
result=df3.corr(method='spearman')
result

1회_평균매출액      방문빈도
1회_평균매출액  1.000000 -0.498841
방문빈도     -0.498841  1.000000

1회 평균 매출액 순위와 방문빈도 순위는 음의 상관관계가 존재한다. 

### 6.4. 명목형 변수의 상관관계분석

#### 클레임 접수여부와 구매유형간의 연관성이 존재하는지 확인해보자. 

- 귀무가설: 구매유형과 클레임 접수 여부는 연관성이 없다. 
- 연구가설: 구매유형과 클레임 접수 여부는 연관성이 있다. 

In [16]:
#빈도 교차표 작성
import numpy as np
dt1=np.array(df.클레임접수여부)
dt2=np.array(df.구매유형)
dt3=pd.crosstab(dt1,dt2,margins=True)
print(dt3)

col_0   1    2    3    4   All
row_0                         
0      24  197   57  263   541
1      19  120   87  233   459
All    43  317  144  496  1000


In [20]:
obs = np.array([[ 24,197,57,263],[19,120,87,233]])

In [23]:
import scipy as sp#카이제곱검정
sp.stats.chi2_contingency(obs)# 검정통계량, 유의확률, 자유도, 예측빈도

(20.76500540576355,
 0.00011779336652059586,
 3,
 array([[ 23.263, 171.497,  77.904, 268.336],
        [ 19.737, 145.503,  66.096, 227.664]]))

#### 결과 해석
- 성별에 따른 할인금액 수준을 비교해 여성이 남성보다 최대 할인금액이 높은 것으로 나타나지만, 최소값, 평균, 그리고 제1사분위, 제2사분위, 제3사분위의 차이가 크지 않은 것 확인
- 성별에 따라 할인쿠폰 사용 수준이 크게 다르다고 볼 수 없으므로 성별에 따른 할인쿠폰 제도 변화는 불필요해 보임.